Allow sync HTTP fetches to be interrupted to fix hanging

grorp opened the pull request:
Fixes #12815, "if I try to start a game while I'm downloading a mod, Minetest hangs", "if I try to close Minetest while I'm downloading a mod, Minetest hangs", etc. This issue is a common cause of ANRs on Android.

This PR is an adoption of #12983. It works by making sync HTTP fetches interruptible if they are not on the main thread. I removed the `FreeThread` thing which was included in the original PR because of a suggestion to make this simpler and it still seems to work, but I have no idea what I'm doing here. Please review this carefully.

Alternative PR: #14411

**Pros of this PR**

- simpler
- allows keeping stuff like JSON decoding, ZIP unpacking, etc. in a worker thread without extra effort

**Cons of this PR**

- I don't know what I'm doing

## To do

This PR is a Ready for Review.

- [x] Decide whether this or the alternative PR is the right approach
- [x] Fix attribution to TurkeyMcMac

## How to test

Start downloading a package from ContentDB and try to close Minetest. Verify that it doesn't hang.

sfan5 commented on src/httpfetch.cpp:
```diff
@@ -678,7 +679,7 @@ class CurlFetchThread : public Thread
 			*/
 			if (m_all_ongoing.empty())
-				waitForRequest(100000000);
+				waitForRequest(5000);
```
why do we need this?

grorp replied:
I unfortunately don't know, undoing the change doesn't seem to make any difference.

sfan5 replied:
This might have interacted with the now removed `FreeThread` but I don't see a reason for it now.
Please change it back.

sfan5 commented on src/httpfetch.cpp:
```diff
@@ -766,6 +767,27 @@ void httpfetch_sync(const HTTPFetchRequest &fetch_request,
+		httpfetch_async(req);
+		do {
+			if (thread->stopRequested()) {
+				httpfetch_caller_free(req.caller);
```
afaict `httpfetch_caller_free` handles all cases correctly so the `FreeThread` from the previous PR was indeed not needed

sfan5 commented on src/threading/thread.cpp:
```diff
@@ -177,6 +180,8 @@ void Thread::threadProc(Thread *thr)
+
 	thr->setName(thr->m_name);
 
 	g_logger.registerThread(thr->m_name);
```
looks like the logger could possibly use the new method too :)

grorp replied:
The logger also handles non-registered threads:

https://github.com/minetest/minetest/blob/cdce33dd05cfc9c13900e5a7cffcc41f8051237b/src/log.cpp#L236-L248

So would it be safe to assume that `Thread::getCurrentThread() == nullptr` means "main thread"? Or are there threads not handled by the `Thread` class?

sfan5 replied:
Not sure. But this also doesn't matter too much, better leave it as-is.

sfan5 commented on src/httpfetch.cpp:
```diff
@@ -766,6 +767,27 @@ void httpfetch_sync(const HTTPFetchRequest &fetch_request,
+bool httpfetch_sync_interruptible(const HTTPFetchRequest &fetch_request,
+		HTTPFetchResult &fetch_result, long interval)
+{
+	if (Thread *thread = Thread::getCurrentThread()) {
```
thought: does clientlauncher/guiEngine even run in a `Thread`? I remember seeing code that used `porting::signal_handler_killstatus` and other custom ways to abort.

(we might have to create a `Thread` object for the main thread...)

grorp replied:
I think not, but all the HTTP requests happen in `core.handle_async` calls, which are `Thread`s, so it doesn't matter in practice (except for the "get CDB package dependencies" request)

sfan5 replied:
oh true, for some reason I forgot about the async thing.

sfan5 commented:
My test code:
```diff
diff --git a/builtin/mainmenu/init.lua b/builtin/mainmenu/init.lua
index 41885e298..cbfd22b95 100644
--- a/builtin/mainmenu/init.lua
+++ b/builtin/mainmenu/init.lua
@@ -127,3 +127,10 @@ local function init_globals()
 end
 
 init_globals()
+
+core.handle_async(function()
+	local x = core.get_http_api().fetch_sync({
+		url = "https://httpbin.org/delay/10",
+	})
+	print("returned " .. dump(x))
+end, nil, function() end)
diff --git a/src/httpfetch.cpp b/src/httpfetch.cpp
index a6b86e98d..ac088f77c 100644
--- a/src/httpfetch.cpp
+++ b/src/httpfetch.cpp
@@ -747,9 +747,11 @@ static void httpfetch_request_clear(u64 caller)
 {
 	if (g_httpfetch_thread->isRunning()) {
 		Event event;
+		rawstream << "request clear + wait" << std::endl;
 		g_httpfetch_thread->requestClear(caller, &event);
 		event.wait();
 	} else {
+		rawstream << "request clear" << std::endl;
 		g_httpfetch_thread->requestClear(caller, nullptr);
 	}
 }
@@ -776,6 +778,7 @@ bool httpfetch_sync_interruptible(const HTTPFetchRequest &fetch_request,
 		httpfetch_async(req);
 		do {
 			if (thread->stopRequested()) {
+				rawstream << "interrupted" << std::endl;
 				httpfetch_caller_free(req.caller);
 				return false;
 			}
@@ -783,6 +786,7 @@ bool httpfetch_sync_interruptible(const HTTPFetchRequest &fetch_request,
 		} while (!httpfetch_async_get(req.caller, fetch_result));
 		httpfetch_caller_free(req.caller);
 	} else {
+		rawstream << "not interruptible!" << std::endl;
 		httpfetch_sync(fetch_request, fetch_result);
 	}
 	return true;
```

sfan5 commented on src/gui/guiEngine.cpp:
```diff
@@ -663,7 +663,8 @@ bool GUIEngine::downloadFile(const std::string &url, const std::string &target)
 		(long)g_settings->getS32("curl_file_download_timeout"));
-	httpfetch_sync(fetch_request, fetch_result);
+	if (!httpfetch_sync_interruptible(fetch_request, fetch_result))
+		return false;
```
should close + delete

sfan5 commented on src/threading/thread.h:
```diff
@@ -119,6 +119,11 @@ class Thread {
 	bool setPriority(int prio);
 
+	/*
+	 * Returns the current thread object if it exists.
```
```suggestion
	 * Returns the thread object of the current thread object if it exists.
```

grorp replied:
```suggestion
	 * Returns the thread object of the current thread if it exists.
```

I guess you meant this?

sfan5 replied:
yes

sfan5 reviewed (approved):
works otherwise

grorp commented:
Fixed attribution to TurkeyMcMac and rebased. Also added this change from the original PR that I apparently had missed: https://github.com/minetest/minetest/pull/12983/commits/d9a52bb18a6e13d8187bf0838aa2067849c64973

nerzhul commented on src/httpfetch.h:
```diff
@@ -135,6 +135,9 @@ u64 httpfetch_caller_alloc_secure();
+// thread is a Thread object. interval is the completion check interval in ms.
+// This blocks and therefore should only be used from background threads.
+// Returned is whether the request completed without interruption.
+bool httpfetch_sync_interruptible(const HTTPFetchRequest &fetch_request,
```
as you removed http_fetch_sync, i think we can call this httpfetch_sync which is the standard and have a default value

grorp replied:
Well, httpfetch_sync still exists, it's just private to httpfetch.cpp now.
I'd have to rename httpfetch_sync to something else. Do you have a suggestion?
In the end, it might be cleaner to keep the names as is.
